### ISSUE TITLE
Add option to warn about unused functions

### ISSUE THREAD

**Issue by mpickering** [haskell/haddock#490]
With the new hyperlinked source infrastructure, it is quite easy to write a script which checks to see whether functions are ever used in a project. For libraries, this isn't that useful as you often define functions which you don't use but for executables it could be very useful to eliminate dead code. 

If anyone is interested in picking this up, there is a hacked together implementation on this branch - https://github.com/mpickering/haddock/tree/ghc-headnew

**Comment by crockeea** [comment]
How is this proposal different from GHC's `-fwarn-unused-binds`?

**Comment by mpickering** [comment]
`-Wunused-binds` doesn't work across modules and a binding is used if exported. My intended use is for executables or end users to eliminate dead code.

**Comment by bgamari** [comment]
Matthew Pickering <notifications@github.com> writes:

> `-Wunused-binds` doesn't work across modules and a binding is used if
> exported. My intended use is for executables or end users to eliminate
> dead code.
>
Indeed this feature sounds incredibly useful. However, I do wonder whether
`haddock` is really the right place for it.

**Comment by mpickering** [comment]
I agree. However, the reason why I opened the ticket here is that I have found that whenever I want to run some kind of analysis then hacking it into haddock is far more reliable than trying to use the GHC API myself. Perhaps this ticket should really be about factoring out the bits of haddock so that consumers can use them to write these kinds of tools. I feel though that haddock is very tightly coupled so it would be quite difficult.
